### PR TITLE
Data Type Options

### PR DESCRIPTION
**Pull request by ghislaineguerin**
Fixes [#1578 ](https://github.com/centerofci/mathesar/issues/1578)

### CODE REVIEW

File: design/specs/data-type-options.md
Comment (pavish) [inline design/specs/data-type-options.md:48]: 1. Could you elaborate on this? I don't understand what is meant by removing sub-sections for database-level options.
2. The image does not show any warning or icon. What would the warning be and how would it look like?

File: design/specs/data-type-options.md
Comment (pavish) [inline design/specs/data-type-options.md:40]: I can see that the gain obtained in making the input labels inline and using a select menu instead of scrollable list would make up for the vertical space lost in getting rid of tabs.

I think we can proceed with these set of changes for now.

File: design/specs/data-type-options.md
Comment (pavish) [inline design/specs/data-type-options.md:50]: For both the proposed solution and the alternative:

We always need an explicit Save button. This save button _needs to be common_ for all the database options combined and _**not** per input_ because the determination of column type depends on the fully filled database options form. The behaviour of the save button and it's placement needs to be mentioned clearly on the spec.

**Review by pavish — CHANGES_REQUESTED** [review]

File: design/specs/data-type-options.md
Comment (ghislaineguerin) [reply design/specs/data-type-options.md:50]: @pavish we have a separate spec for inline saving. I will include a reference here and have a wireframe for reference.

Comment (pavish) [reply design/specs/data-type-options.md:50]: I understand we have a separate spec for inline saving, my question is whether it will cover the entire form?

> we have a separate spec for inline saving

The current inline saving seems to be per input, each input having it's own save button. But for the data type menu, we cannot do that, because the type is calculated based on the result of all the related inputs. So, we need a common save button.

File: design/specs/data-type-options.md
Comment (ghislaineguerin) [reply design/specs/data-type-options.md:48]: 1. I meant we don't need to hace sub-headings for "Database" and "Format" options.
2. I will include an example.

File: design/specs/data-type-options.md
Comment (ghislaineguerin) [reply design/specs/data-type-options.md:50]: @pavish yes, the entire section should be saved at once. I will include the example in the spec.

Comment (ghislaineguerin) [reply design/specs/data-type-options.md:46]: @pavish I'm assuming that save is only required for the database options and that we won't be asking the user to save format options. Is that the right assumption?

Comment (pavish) [reply design/specs/data-type-options.md:46]: Yes, we can make format options save automatically. However, **I'd prefer the save button to save both database and format options together**. We show the same options in both data import menu and the table inspector, and it would be a lot simpler implementation-wise and maybe more user friendly, if they were as similar to each other as possible.

Comment (ghislaineguerin) [reply design/specs/data-type-options.md:46]: It makes sense. I was concerned about the inconsistency between this and inline saving but I think you make a good point. I will update.

**Comment by ghislaineguerin** [comment]
@pavish I have updated the wireframe to show the save and cancel options for the entire section.

**Review by kgodey — COMMENTED** [review]
This looks fine to me, but default value should not be folded into the Data Type "save" action, it should be a separate action.

Also @ghislaineguerin please remove unused images from the PR.

I think we generally need to do some explanatory copy for the table inspector, but we can handle that separately as part of demo video readiness.

**Review by pavish — CHANGES_REQUESTED** [review]
I find the UX a bit confusing because Database options includes the "Data type" Select menu, and it contains a border. The Formatting options are placed out of it, even though they depend on the selected data type.

I think the "Data Type" menu should be placed out of the Database options border, or we should find a way where it is clear that the formatting options depend on the selected data type.

I also agree with Kriti that the default value option should not be folded with the "save" action.

I'd prefer it if 'Default value' is not part of the Data type section, but placed within a constraint section, but this is something we can come back to later.

File: design/specs/data-type-options.md
Comment (pavish) [reply design/specs/data-type-options.md]: I have further comments placed here: https://github.com/centerofci/mathesar-wiki/pull/59#pullrequestreview-1098147017

**Comment by ghislaineguerin** [comment]
@pavish The default options are a separate section. I agree with your concern about the border, I wasn't sure if grouping the data type selection was a good idea, but I did since it was technically a database option as well. But it might be better as a separate option that determines all the others. I will update.

**Review by pavish — COMMENTED** [review]
@ghislaineguerin @kgodey

The spec looks good to me.

I do have one concern, which didn't occur to me earlier.

* Whenever we change the type of a column, the default value has to change/get cleared.
  * Since the default value is outside the database form, it may not be immediately known to the user that it has been cleared.
  * I personally find it okay, and I think we can show a confirmation alert modal before changing the type of a column (same as the one we would have for deleting), since column type change would also have data loss implications. We can show those warnings in that modal. 
* Or, we can retain the default value and let the user modify it manually. The column type change would mostly throw an error if the user tries to update the column without removing the default value first.

Which option should we go with? It's not immediately blocking, I've already started implementation on this.

**Comment by kgodey** [comment]
>  I think we can show a confirmation alert modal before changing the type of a column

I think this is fine, can we change the wording of the modal if there's a default value set? We can indicate that the default value will be cleared.

**Comment by pavish** [comment]
> can we change the wording of the modal if there's a default value set

Yes, we can do that.

**Review by pavish — APPROVED** [review]
@ghislaineguerin I've added a commit https://github.com/centerofci/mathesar-wiki/pull/59/commits/ffea0d1fb0fc8b1b5db798b8dec6e0bd4f54f2d8 addressing https://github.com/centerofci/mathesar-wiki/pull/59#pullrequestreview-1104535767, https://github.com/centerofci/mathesar-wiki/pull/59#issuecomment-1244085603.

I'm approving the PR from my end. I've leave it to you to take a look before merging.

**Pull request merged**
